src/http.cc

```diff
@@ -833,11 +846,112 @@ HttpStateData::handle1xx(HttpReply *reply)
+{
+    const auto acceptedProtos = reply.header.getList(Http::HdrType::UPGRADE);
+
+    if (!upgradeHeaderOut) {
+        debugs(11, 2, "Squid offered no Upgrade at all, but server accepted " << acceptedProtos);
+        return false;
+    }
+
+    for(const auto &accepted: StrList(acceptedProtos)) {
+        const ProtocolView acceptedProto(accepted);
+        for (const auto &offered: StrList(*upgradeHeaderOut)) {
+            const ProtocolView offeredProto(offered);
+            if (Similar(offeredProto, acceptedProto))
+                return true;
+        }
+        debugs(11, 2, "Squid did not offer, but server accepted " << acceptedProto);
+        return false;
+    }
+
+    debugs(11, 2, "server accepted no Upgrade protocols: " << acceptedProtos);
+    return false;
+}
+
+/// \retval nil if the HTTP/101 (Switching Protocols) reply should be forwarded
+/// \retval reason why an attempt to switch protocols should be stopped
+const char *
+HttpStateData::blockSwitchingProtocols(const HttpReply &reply) const
+{
+    // See RFC 7230 section 6.7 for the corresponding MUSTs
+
+    if (!reply.header.has(Http::HdrType::UPGRADE))
+        return "server did not send an Upgrade header field";
+
+    if (!reply.header.hasListMember(Http::HdrType::CONNECTION, "upgrade", ','))
+        return "server did not send 'Connection: upgrade'";
+
+    if (!serverSwitchedToOfferedProtocols(reply))
+        return "server switched to unsolicited protocol(s)";
+
+    return nullptr; // no Upgrade violations detected
+}
+
+// XXX: To control ownership, use a custom unique_ptr instead of a custom Dialer
+/// Passes server connection ownership to the ConnStateData object.
+/// Closes the connection if ConnStateData disappears.
+class ServerConnectionControlDialer: public UnaryMemFunT<ConnStateData, ConnStateData::ServerConnectionContext>
+{
+public:
+    typedef UnaryMemFunT<ConnStateData, ConnStateData::ServerConnectionContext> Parent;
+
+    ServerConnectionControlDialer(const CbcPointer<ConnStateData> &mgr, const ConnStateData::ServerConnectionContext &scc): Parent(mgr, &ConnStateData::noteTakeServerConnectionControl, scc) {}
+
+    virtual bool canDial(AsyncCall &call) override {
+        if (!Parent::canDial(call)) {
+            debugs(11, 2, "ConnStateData gone, close server connection to avoid orphan connection!");
+            arg1.connection->close();
+            return false;
+        }
+        return true;
+    }
+
+    const char *callName = "ConnStateData::noteTakeServerConnectionControl";
+};
+
 /// restores state and resumes processing after 1xx is ignored or forwarded
 void
 HttpStateData::proceedAfter1xx()
 {
     Must(flags.handling1xx);
+
+    if (flags.serverSwitchedProtocols) {
+        // pass server connection ownership to request->clientConnectionManager
+        ConnStateData::ServerConnectionContext scc(serverConnection, request);
+        ServerConnectionControlDialer dialer(request->clientConnectionManager, scc);
```

Comment:
I do not remember any more.
But I have passed many phases before conclude to this design decision.
I am not sure about this now, but if I am not wrong if the Http server did not get any client data after respond with 101 after a timeout, it may proceeds with HTTP protocol. OR something like that, I need to check again.